### ISSUE TITLE
Issues with SMILES Encoding and Decoding in Common Cases Using the SMI-TED Model

### ISSUE THREAD

**Issue by yueyue2299** [IBM/materials#20]
I was surprised to observe that the SMI-TED model produces incorrect predictions in some common cases during SMILES encoding and decoding. For example:

C → Brc1
CC(C)C(C)C → CCC)C(C)C
Cl → Clc1
F → Fc1

The provided data file, moses_test.csv, seems to be perfectly formatted and does not appear to contain any issues.

Could you help clarify if this behavior is expected or if there might be a potential bug? Please let me know if additional details or examples would be helpful. Thank you!

**Comment by vichShir** [comment]
Hi! Thanks for your observation. With the open-sourced checkpoint, we observed a reconstruction of 99.09% (4520667/4562412) on a random sample of 4M SMILES in our pretraining data, but it's expected to SMI-TED model to not reconstruct some cases. The reasons for misclassified tokens may vary, but we observed more in SMILES with single token, rare tokens, lengthy SMILES. Since our Decoder predicts `MAX_LEN` tokens (the `MAX_LEN` for SMI-TED-Light is 202 tokens) once per SMILES, it may incorrectly predict some tokens due to class imbalance.

Just to illustrate this, let's try to reconstruct the following 4 SMILES:
- `C`
- `[Mg].[Ni]`
- `CO`
- `Brc1ccc2c(c1)CC(C1CCCC1)=N2.CC(=O)C(C)(C)C.CC(=O)N1CCCC1C1=Nc2ccc(Br)cc2C1.CC(C)(C)OC(=O)N1CCCC1C1=Nc2ccc(Br)cc2C1.CC(C)C.CN1CCCC1C1=Nc2ccc(B3OC(C)(C)C(C)(C)O3)cc2C1.IC(I)I.ICI.II.I[V]I.S.S.S.S`:
> **Special tokens**
> Begin token: 0
> End token: 1
> Padding token: 2

#### Incorrect reconstructions:
```
INPUT: smi_ted('C')
TOKENIZATION: tensor([[0, 4, 1]])
PREDICTED TOKENS: [[ 0 27  5  8  1  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
   2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
   2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
   2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
   2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
   2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
   2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
   2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
   2  2  2  2  2  2  2  2  2  2]]
RECONSTRUCTION: Brc1
```

```
INPUT: smi_ted('[Mg].[Ni]')
TOKENIZATION: tensor([[  0, 197,  34, 129,   1]])
PREDICTED TOKENS: [[  0 129  34 129   1   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    2   2   2   2]]
RECONSTRUCTION: [Ni].[Ni]
```

#### Correct reconstructions:
```
INPUT: smi_ted('CO')
TOKENIZATION: tensor([[0, 4, 9, 1]])
PREDICTED TOKENS: [[0 4 9 1 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
  2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]]
RECONSTRUCTION: CO
```

```
INPUT: smi_ted('Brc1ccc2c(c1)CC(C1CCCC1)=N2.CC(=O)C(C)(C)C.CC(=O)N1CCCC1C1=Nc2ccc(Br)cc2C1.CC(C)(C)OC(=O)N1CCCC1C1=Nc2ccc(Br)cc2C1.CC(C)C.CN1CCCC1C1=Nc2ccc(B3OC(C)(C)C(C)(C)O3)cc2C1.IC(I)I.ICI.II.I[V]I.S.S.S.S')
TOKENIZATION: tensor([[  0,  27,   5,   8,   5,   5,   5,  11,   5,   6,   5,   8,   7,   4,
           4,   6,   4,   8,   4,   4,   4,   4,   8,   7,  12,  10,  11,  34,
           4,   4,   6,  12,   9,   7,   4,   6,   4,   7,   6,   4,   7,   4,
          34,   4,   4,   6,  12,   9,   7,  10,   8,   4,   4,   4,   4,   8,
           4,   8,  12,  10,   5,  11,   5,   5,   5,   6,  27,   7,   5,   5,
          11,   4,   8,  34,   4,   4,   6,   4,   7,   6,   4,   7,   9,   4,
           6,  12,   9,   7,  10,   8,   4,   4,   4,   4,   8,   4,   8,  12,
          10,   5,  11,   5,   5,   5,   6,  27,   7,   5,   5,  11,   4,   8,
          34,   4,   4,   6,   4,   7,   4,  34,   4,  10,   8,   4,   4,   4,
           4,   8,   4,   8,  12,  10,   5,  11,   5,   5,   5,   6,  47,  14,
           9,   4,   6,   4,   7,   6,   4,   7,   4,   6,   4,   7,   6,   4,
           7,   9,  14,   7,   5,   5,  11,   4,   8,  34,  35,   4,   6,  35,
           7,  35,  34,  35,   4,  35,  34,  35,  35,  34,  35, 122,  35,  34,
          18,  34,  18,  34,  18,  34,  18,   1]])
PREDICTED TOKENS: [[  0  27   5   8   5   5   5  11   5   6   5   8   7   4   4   6   4   8
    4   4   4   4   8   7  12  10  11  34   4   4   6  12   9   7   4   6
    4   7   6   4   7   4  34   4   4   6  12   9   7  10   8   4   4   4
    4   8   4   8  12  10   5  11   5   5   5   6  27   7   5   5  11   4
    8  34   4   4   6   4   7   6   4   7   9   4   6  12   9   7  10   8
    4   4   4   4   8   4   8  12  10   5  11   5   5   5   6  27   7   5
    5  11   4   8  34   4   4   6   4   7   4  34   4  10   8   4   4   4
    4   8   4   8  12  10   5  11   5   5   5   6  47  14   9   4   6   4
    7   6   4   7   4   6   4   7   6   4   7   9  14   7   5   5  11   4
    8  34  35   4   6  35   7  35  34  35   4  35  34  35  35  34  35 122
   35  34  18  34  18  34  18  34  18   1   2   2   2   2   2   2   2   2
    2   2   2   2]]
RECONSTRUCTION: Brc1ccc2c(c1)CC(C1CCCC1)=N2.CC(=O)C(C)(C)C.CC(=O)N1CCCC1C1=Nc2ccc(Br)cc2C1.CC(C)(C)OC(=O)N1CCCC1C1=Nc2ccc(Br)cc2C1.CC(C)C.CN1CCCC1C1=Nc2ccc(B3OC(C)(C)C(C)(C)O3)cc2C1.IC(I)I.ICI.II.I[V]I.S.S.S.S
```

One may try to fine-tune the Decoder, so it'll improve its reconstructions. You can use [this training code](https://github.com/IBM/materials/blob/main/models/smi_ted/training/train_model_D.py) to only adjust the Decoder weights. Let me know if you have more questions.

**Comment by yueyue2299** [comment]
Thanks a lot for your detailed explanation and for taking the time to test the model! Your insights about the tokenization and reconstruction behavior really clarified my doubts. I truly appreciate your effort in helping me understand this better.

**Comment by vichShir** [comment]
You're welcome! 

I'll be closing this issue. Please, feel free to reopen if it has been closed in error.
